VideoLocker/src/main/java/org/edx/mobile/view/DownloadListActivity.java

```diff
@@ -56,68 +38,57 @@ public void onItemClicked(DownloadEntry model) {
             }
 
             @Override
-            public void onDownloadComplete(DownloadEntry model) {
-                //showDownloadCompleteView();
-                adapter.remove(model);
-                adapter.notifyDataSetChanged();
-            }
-
-            @Override
             public void onDeleteClicked(DownloadEntry model) {
-                if(environment.getStorage().removeDownload(model) >= 1){
+                if (environment.getStorage().removeDownload(model) >= 1) {
                     // update the list data as one download is removed
-                    try{
-                        adapter.remove(model);
-                        adapter.notifyDataSetChanged();
-                    }catch(Exception e){
-                        logger.error(e);
-                    }
-
+                    adapter.remove(model);
+                    adapter.notifyDataSetChanged();
                 }
             }
         };
-
-        adapter.setStore(environment.getStorage());
+        final ListView downloadListView = (ListView) findViewById(R.id.my_downloads_list);
         downloadListView.setAdapter(adapter);
-        final ArrayList<DownloadEntry> list = new ArrayList<DownloadEntry>();
+    }
+
+    private void observeOngoingDownloads() {
         environment.getDatabase().getListOfOngoingDownloads(new DataCallback<List<VideoModel>>() {
             @Override
-            public void onResult(List<VideoModel> result) {
-                if(result!=null){
-                    for(VideoModel de : result){
-                        list.add((DownloadEntry) de);
-                    }
-                    adapter.setItems(list);
+            public void onResult(final List<VideoModel> result) {
+                if (result != null) {
+                    handler.post(new Runnable() {
+                        @Override
+                        public void run() {
+                            adapter.setItems((List) result);
+                        }
+                    });
                 }
             }
+
             @Override
             public void onFail(Exception ex) {
                 logger.error(ex);
             }
         });
+        handler.postDelayed(new Runnable() {
+            @Override
+            public void run() {
+                observeOngoingDownloads();
```

Comment:
> I just realized this also leaks the activity context for the duration of `getListOfOngoingDownloads`.

That's currently an issue for _all_ of our asynchronous callbacks (HTTP calls as well as database queries). Switching to a `Loader` implementation would fix that, amongst other issues :wink:

> There is also a DB hit per list item being rendered, which causes noticable lag when done on the UI thread.

Yeah, that's another general issue that's present in `CourseOutlineAdapter` as well. We should implement a generic optimization for this at some point.